File: shell.go
package main

import (
	"fmt"
	"os"
	"strings"
	"syscall"
	"unicode"
	"unsafe"
)

func shellInit() {
	sh.ShellPid = os.Getpid()
	sh.LastStatus = 0
	sh.NextJobId = 1
	sh.Vars = make(map[string]Var)
	sh.Aliases = make(map[string]string)
	sh.Functions = make(map[string]*FuncDef)
	sh.Trap = make(map[string]string)

	sh.Vars["HOME"] = Var{Value: getEnv("HOME", ""), Flags: 0}
	sh.Vars["PATH"] = Var{Value: getEnv("PATH", "/usr/local/bin:/usr/bin:/bin"), Flags: VAR_EXPORT}
	sh.Vars["USER"] = Var{Value: getEnv("USER", "meow"), Flags: 0}
	sh.Vars["PWD"] = Var{Value: getEnv("PWD", "/"), Flags: 0}
	sh.Vars["SHELL"] = Var{Value: "/bin/meowsh", Flags: 0}
	sh.Vars["TERM"] = Var{Value: getEnv("TERM", "dumb"), Flags: 0}
	sh.Vars["PS1"] = Var{Value: "𓃠 ", Flags: 0}
	sh.Vars["PS2"] = Var{Value: "𓃠 ", Flags: 0}
	sh.Vars["SHLVL"] = Var{Value: "1", Flags: 0}

	jobsInit()
}

func getEnv(key, def string) string {
	if val := os.Getenv(key); val != "" {
		return val
	}
	return def
}

func varGet(name string) string {
	switch name {
	case "$":
		return fmt.Sprintf("%d", sh.ShellPid)
	case "?":
		return fmt.Sprintf("%d", sh.LastStatus)
	case "!":
		return fmt.Sprintf("%d", sh.LastBgPid)
	case "0":
		return sh.Argv1()
	case "#":
		return fmt.Sprintf("%d", len(sh.PosParams))
	case "@", "*":
		return strings.Join(sh.PosParams, " ")
	case "-":
		flags := ""
		if sh.Opts&OPT_ALLEXPORT != 0 {
			flags += "a"
		}
		if sh.Opts&OPT_ERREXIT != 0 {
			flags += "e"
		}
		if sh.Opts&OPT_NOGLOB != 0 {
			flags += "f"
		}
		if sh.Opts&OPT_HASHALL != 0 {
			flags += "h"
		}
		if sh.Opts&OPT_INTERACTIVE != 0 {
			flags += "i"
		}
		if sh.Opts&OPT_MONITOR != 0 {
			flags += "m"
		}
		if sh.Opts&OPT_NOEXEC != 0 {
			flags += "n"
		}
		if sh.Opts&OPT_NOUNSET != 0 {
			flags += "u"
		}
		if sh.Opts&OPT_VERBOSE != 0 {
			flags += "v"
		}
		if sh.Opts&OPT_XTRACE != 0 {
			flags += "x"
		}
		return flags
	}

	if len(name) > 1 && name[0] == '-' {
		return varGet(name[1:])
	}

	if v, ok := sh.Vars[name]; ok {
		return v.Value
	}
	return ""
}

func varSet(name, value string, export bool) {
	flags := sh.Vars[name].Flags
	if export {
		flags |= VAR_EXPORT
	}
	sh.Vars[name] = Var{Value: value, Flags: flags}
	os.Setenv(name, value)
}

func varUnset(name string) error {
	if v, ok := sh.Vars[name]; ok {
		if v.Flags&VAR_READONLY != 0 {
			return fmt.Errorf("cannot unset readonly variable %s", name)
		}
	}
	delete(sh.Vars, name)
	os.Unsetenv(name)
	return nil
}

func (s *Shell) Argv1() string {
	if len(s.PosParams) > 0 {
		return s.PosParams[0]
	}
	return s.Argv0
}

func shortenPath(p string) string {
	home := varGet("HOME")
	if home != "" && strings.HasPrefix(p, home) {
		return "~" + strings.TrimPrefix(p, home)
	}

	parts := strings.Split(p, "/")
	if len(parts) <= 3 {
		return p
	}

	var result []string
	for i, part := range parts {
		if i == 0 && part == "" {
			result = append(result, "/")
		} else if i == len(parts)-1 {
			result = append(result, part)
		} else if part != "" {
			result = append(result, string(part[0]))
		}
	}
	return strings.Join(result, "/")
}

func buildPrompt() string {
	user := varGet("USER")
	if user == "" {
		user = "meow"
	}
	pwd := varGet("PWD")
	if pwd == "" {
		pwd = "?"
	}

	shortPwd := shortenPath(pwd)

	return fmt.Sprintf("\033[32m%s\033[0m \033[34m%s\033[0m 𓃠  ", user, shortPwd)
}

func isatty(fd uintptr) bool {
	var termios syscall.Termios
	_, _, err := syscall.Syscall(syscall.SYS_IOCTL, fd, uintptr(syscall.TCGETS), uintptr(unsafe.Pointer(&termios)))
	return err == 0
}

func isAssignment(s string) bool {
	for i, c := range s {
		if c == '=' {
			return i > 0
		}
		if !unicode.IsLetter(c) && !unicode.IsDigit(c) && c != '_' {
			return false
		}
	}
	return false
}

File: meowsh_rs/src/shell.rs
use crate::types::{
    Shell, Var, OPT_ALLEXPORT, OPT_ERREXIT, OPT_HASHALL, OPT_INTERACTIVE, OPT_MONITOR, OPT_NOEXEC,
    OPT_NOGLOB, OPT_NOUNSET, OPT_VERBOSE, OPT_XTRACE, VAR_EXPORT,
};
use std::env;
use std::path::PathBuf;
use std::sync::Mutex;

pub struct ShellState {
    pub shell: Mutex<Shell>,
}

impl Default for ShellState {
    fn default() -> Self {
        Self::new()
    }
}

impl ShellState {
    pub fn new() -> Self {
        ShellState {
            shell: Mutex::new(Shell::new()),
        }
    }
}

lazy_static::lazy_static! {
    pub static ref SHELL: ShellState = ShellState::new();
}

pub fn shell_init() {
    let mut shell = SHELL.shell.lock().unwrap();
    shell.shell_pid = std::process::id() as i32;
    shell.last_status = 0;
    *shell.next_job_id.lock().unwrap() = 1;

    shell.vars.insert(
        "HOME".to_string(),
        Var {
            value: get_env("HOME", ""),
            flags: 0,
        },
    );
    shell.vars.insert(
        "PATH".to_string(),
        Var {
            value: get_env("PATH", "/usr/local/bin:/usr/bin:/bin"),
            flags: VAR_EXPORT,
        },
    );
    shell.vars.insert(
        "USER".to_string(),
        Var {
            value: get_env("USER", "meow"),
            flags: 0,
        },
    );
    shell.vars.insert(
        "PWD".to_string(),
        Var {
            value: get_env("PWD", "/"),
            flags: 0,
        },
    );
    shell.vars.insert(
        "SHELL".to_string(),
        Var {
            value: "/bin/meowsh".to_string(),
            flags: 0,
        },
    );
    shell.vars.insert(
        "TERM".to_string(),
        Var {
            value: get_env("TERM", "dumb"),
            flags: 0,
        },
    );
    shell.vars.insert(
        "PS1".to_string(),
        Var {
            value: "meowsh> ".to_string(),
            flags: 0,
        },
    );
    shell.vars.insert(
        "PS2".to_string(),
        Var {
            value: "meowsh... ".to_string(),
            flags: 0,
        },
    );
    shell.vars.insert(
        "SHLVL".to_string(),
        Var {
            value: "1".to_string(),
            flags: 0,
        },
    );

    if let Some(home) = shell.vars.get("HOME") {
        shell.history_file = PathBuf::from(&home.value)
            .join(".meowsh_history")
            .to_string_lossy()
            .to_string();
    }
}

fn get_env(key: &str, def: &str) -> String {
    env::var(key).unwrap_or_else(|_| def.to_string())
}

pub fn var_get(name: &str) -> String {
    let shell = SHELL.shell.lock().unwrap();
    match name {
        "$" => return format!("{}", shell.shell_pid),
        "?" => return format!("{}", shell.last_status),
        "!" => return format!("{}", shell.last_bg_pid),
        "0" => return shell.argv0.clone(),
        "#" => return format!("{}", shell.pos_params.len()),
        "@" | "*" => return shell.pos_params.join(" "),
        "-" => {
            let mut flags = String::new();
            if shell.opts & OPT_ALLEXPORT != 0 {
                flags.push('a');
            }
            if shell.opts & OPT_ERREXIT != 0 {
                flags.push('e');
            }
            if shell.opts & OPT_NOGLOB != 0 {
                flags.push('f');
            }
            if shell.opts & OPT_HASHALL != 0 {
                flags.push('h');
            }
            if shell.opts & OPT_INTERACTIVE != 0 {
                flags.push('i');
            }
            if shell.opts & OPT_MONITOR != 0 {
                flags.push('m');
            }
            if shell.opts & OPT_NOEXEC != 0 {
                flags.push('n');
            }
            if shell.opts & OPT_NOUNSET != 0 {
                flags.push('u');
            }
            if shell.opts & OPT_VERBOSE != 0 {
                flags.push('v');
            }
            if shell.opts & OPT_XTRACE != 0 {
                flags.push('x');
            }
            return flags;
        }
        _ => {}
    }

    if name.len() > 1 && name.starts_with('-') {
        return var_get(&name[1..]);
    }

    if let Some(v) = shell.vars.get(name) {
        return v.value.clone();
    }
    String::new()
}

pub fn var_get_if_exists(name: &str) -> Option<i64> {
    let shell = SHELL.shell.lock().unwrap();
    if let Some(v) = shell.vars.get(name) {
        if let Ok(n) = v.value.parse::<i64>() {
            return Some(n);
        }
    }
    None
}

pub fn var_set(name: &str, value: &str, _export: bool) {
    let mut shell = SHELL.shell.lock().unwrap();
    shell.vars.insert(
        name.to_string(),
        Var {
            value: value.to_string(),
            flags: 0,
        },
    );
    env::set_var(name, value);
}

pub fn is_assignment(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    for (i, c) in s.chars().enumerate() {
        if c == '=' {
            return i > 0;
        }
        if !c.is_alphabetic() && !c.is_ascii_digit() && c != '_' {
            return false;
        }
    }
    false
}

pub fn shorten_path(p: &str) -> String {
    let home = var_get("HOME");
    if !home.is_empty() && p.starts_with(&home) {
        return format!("~{}", p.strip_prefix(&home).unwrap_or(""));
    }
    p.to_string()
}

pub fn build_prompt() -> String {
    let user = var_get("USER");
    let user = if user.is_empty() { "meow" } else { &user };
    let pwd = var_get("PWD");
    let pwd = if pwd.is_empty() { "?" } else { &pwd };
    let short_pwd = shorten_path(pwd);
    format!("\x1b[32m{}\x1b[0m \x1b[34m{}\x1b[0m 𓃠  ", user, short_pwd)
}
